src/main/flight/pid.c

```diff
@@ -291,6 +291,20 @@ void pidResetIterm(void)
     }
 }
 
+void pidUpdateTpaFactor(float throttle)
+{
+    pidRuntime.tpaFactor = 1.0f; 
+    const float tpaBreakpoint = (currentControlRateProfile->tpa_breakpoint - 1000) / 1000.0f;
+    if (throttle > tpaBreakpoint) {
+        const float tpaRate = currentControlRateProfile->tpa_rate / 100.0f;
+        if (throttle < 1.0f) {
+            pidRuntime.tpaFactor -= tpaRate * (throttle - tpaBreakpoint) / (1.0f - tpaBreakpoint);
+        } else {
+            pidRuntime.tpaFactor -= tpaRate;
```

Comment:
This had me scratching my head for a second. For readability I think it should be
```pidRuntime.tpaFactor = 1.0f - (tpaRate * (throttle - tpaBreakpoint) / (1.0f - tpaBreakpoint));```
```pidRuntime.tpaFactor = 1.0f - tpaRate;```

and default to 1 with
```
else { 
    pidRuntime.tpaFactor = 1.0f;
}
```

Reply:
Yes, `-=` is always a bit tricky.
You mean so it would look like this:
```
void pidUpdateTpaFactor(float throttle)
{
    const float tpaBreakpoint = (currentControlRateProfile->tpa_breakpoint - 1000) / 1000.0f;
    if (throttle > tpaBreakpoint) {
        const float tpaRate = currentControlRateProfile->tpa_rate / 100.0f;
        if (throttle < 1.0f) {
            pidRuntime.tpaFactor = 1.0f - tpaRate * (throttle - tpaBreakpoint) / (1.0f - tpaBreakpoint);
        } else {
            pidRuntime.tpaFactor = 1.0f - tpaRate;
        }
    } else {
        pidRuntime.tpaFactor = 1.0f;
    }
}
```

Reply:
The other alternative I came up with was:
```void pidUpdateTpaFactor(float throttle)
{
    const float tpaBreakpoint = (currentControlRateProfile->tpa_breakpoint - 1000) / 1000.0f;
    float tpaRate = currentControlRateProfile->tpa_rate / 100.0f;
    if (throttle > tpaBreakpoint) {
        if (throttle < 1.0f) {
            tpaRate *= (throttle - tpaBreakpoint) / (1.0f - tpaBreakpoint);
        }
    } else {
        tpaRate = 0.0f;
    }
    pidRuntime.tpaFactor = 1.0f - tpaRate;
}
```

Reply:
What is the reason for the `throttle < 1.0f` condition? Would it be a problem to do something like this:
```c
void pidUpdateTpaFactor(float throttle)
{
    const float tpaBreakpoint = (currentControlRateProfile->tpa_breakpoint - 1000) / 1000.0f;
    if (throttle > tpaBreakpoint) {
        const float tpaRate = (currentControlRateProfile->tpa_rate / 100.0f);
        pidRuntime.tpaFactor -= tpaRate * (throttle - tpaBreakpoint) / (1.0f - tpaBreakpoint);
    } else {
        pidRuntime.tpaFactor = 1.0f;
    }
}
```

Reply:
@klutvott123 Could I ask if we should be using pid_init.c to initialise `tpaBreakpoint` and `tpaRate`?  
They are rate variables, and can change at any time in flight, so I had thought that we cannot use pid_init.c for this purpose.

Reply:
@ctzsnooze Either of those are fine I think. First suggestion is more readable IMO 🙂 

> @klutvott123 Could I ask if we should be using pid_init.c to initialise `tpaBreakpoint` and `tpaRate`? They are rate variables, and can change at any time in flight, so I figured that we cannot use pid_init.c for this purpose.

I think it's fine as it is now and I don't think pid_init is the place for it

Reply:
OK thanks I just updated to the second version, which isn't so great to read, but saves a lot on flash, and isn't more CPU intensive.
Also changed two divides to multiplies for efficiency.
No more changes planned from me now.

Reply:
The divides compile same output as the multiplication, so if just for efficiency I think division reads better.

Reply:
With Stepan's help we tested compilation using a 'compiler explorer' and our existing compiler settings.  
The assembler instruction used divides unless I specified a multiply.  Flash usage was 4 bytes less with multiply also.
How can we answer this question with certainty, given our compiler settings?

Reply:
I used compiler explorer as well https://godbolt.org/z/Esv94KK1n.
I can check the actual output with objdump to be sure.

Reply:
Sorry, you are right.  
We use the -ffast-math compiler option, it does replace `/ 1000.0f` with multiply. 
 I'll revert to `/ 1000.0f` format.

Reply:
For STM32F405 output is the same except 4 bytes are added to the `<kAlphabet>` section. I don't know why but it does not impact efficiency.

Reply:
The git commit short sha is compiled into the binary as revision info, but for some weird reason the old revision was still part of my second build. After during a clean re-build both compile to the same size.